### PR TITLE
frequency parameters traffic control

### PR DESCRIPTION
**Pull request by louyuting**
### Describe what this PR does / why we need it
Frequency parameters traffic control implementation.

### Describe how you did it
Supporting reject directly strategy and throttling strategy.
Using token counter to record statistics。

### Special notes for reviews

### CODE REVIEW

File: core/stat/stat_slot_callback.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+var (
+	// global StatSlot callback
+	entryCallBackMap  = make(map[CallbackKey][]StatSlotEntryCallback)
+	exitCallBackMap   = make(map[CallbackKey][]StatSlotExitCallback)
+	callbackUpdateMux = new(sync.RWMutex)
+)
+
+type StatSlotEntryCallback interface {
```
Comment (sczyh30) [inline core/stat/stat_slot_callback.go:21]: Maybe the stat slot callback is not needed? Just make it as another `StatSlot` and add it to slot chain is okay :)

Comment (louyuting) [reply core/stat/stat_slot_callback.go:21]: Adding a new StatSlot to record the concurrency statistic seems make more sense. I will reimplement related logic.

File: core/freq_params_traffic/traffic_shaping.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+		lastPassTime := atomic.LoadInt64(lastPassTimePtr)
+		// calculate the expected pass time
+		expectedTime := lastPassTime + intervalCostTime
+
+		if expectedTime <= currentTimeInMs || expectedTime-currentTimeInMs < c.maxQueueingTimeMs {
+			if atomic.CompareAndSwapInt64(lastPassTimePtr, lastPassTime, currentTimeInMs) {
+				awaitTime := expectedTime - currentTimeInMs
+				if awaitTime > 0 {
+					atomic.StoreInt64(lastPassTimePtr, expectedTime)
+					time.Sleep(time.Duration(awaitTime) * time.Millisecond)
```
Comment (sczyh30) [inline core/freq_params_traffic/traffic_shaping.go:220]: Maybe it's better to use `NewTokenResultShouldWait` instead?

File: core/freq_params_traffic/concurrency_stat_slot.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		if tc == nil {
+			logger.Warnf("Nil traffic controller was found, res: %s", res)
+			continue
+		}
+		idx := tc.getParamIndex()
+		if idx < 0 {
+			logger.Warnf("The TrafficShapingController's paramIndex(%d) less than 0.", idx)
+		}
+		if len(args) <= idx {
+			logger.Warnf("The index of rule is not existed in args. The TrafficShapingController's paramIndex is %d, args: %+v, tc: %+v", idx, args, tc)
```
Comment (sczyh30) [inline core/freq_params_traffic/concurrency_stat_slot.go:25]: Ditto for logging level. Also typo: is not existed -> does not exist

File: core/freq_params_traffic/params_metric.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+// paramsMetric is used for pair <resource, trafficShapingController>.
+type paramsMetric struct {
+	// cache's key is the hot value
+	// cache's value is the counter
+	// ruleTimeCounter record the last add token time of value
+	ruleTimeCounter cache.ConcurrentCounterCache
+	// ruleTokenCounter record the number of token of value
+	ruleTokenCounter cache.ConcurrentCounterCache
+	// goroutineCounter record the number of goroutine of value
+	goroutineCounter cache.ConcurrentCounterCache
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/freq_params_traffic/params_metric.go:21]: make sense

File: core/freq_params_traffic/rule.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
+package freq_params_traffic
+
+import "strconv"
+
+type ControlStrategy int8
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/freq_params_traffic/rule.go:5]: That's ok. Keep the name convention.

File: core/freq_params_traffic/rule.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+const (
+	RejectFaster ControlStrategy = iota
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/freq_params_traffic/rule.go:8]: I want to express the meaning of quick and direct fail.
”Reject“ may be enough.

File: core/freq_params_traffic/rule.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+const (
+	kindInt Kind = 1 + iota
+	kindString
+	KindSum
+)
+
+type SpecificValue struct {
+	ValType  Kind
+	ValueStr string
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/freq_params_traffic/rule.go:31]: Okay

File: core/freq_params_traffic/rule.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+type MetricType int8
+
+const (
+	// Concurrency represents concurrency count.
+	Concurrency MetricType = iota
+	// QPS represents request count per second.
+	QPS
+)
+
+type Kind int
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/freq_params_traffic/rule.go:21]: Here i refer to the data type definition of go standard reflect package.
[https://github.com/golang/go/blob/98d20fb23551a7ab900fcfe9d25fd9cb6a98a07f/src/reflect/type.go#L230](https://github.com/golang/go/blob/98d20fb23551a7ab900fcfe9d25fd9cb6a98a07f/src/reflect/type.go#L230)
I think keep name consistent with standard package is ok.

File: core/freq_params_traffic/slot.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+func (s *FreqPramsTrafficSlot) Check(ctx *base.EntryContext) *base.TokenResult {
+
+	res := ctx.Resource.Name()
+	args := ctx.Input.Args
+	acquire := int64(ctx.Input.AcquireCount)
+	tcs := getTrafficControllerListFor(res)
+
+	for _, tc := range tcs {
+		if tc == nil {
+			logger.Warnf("Nil traffic controller was found, res: %s", res)
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/freq_params_traffic/slot.go:20]: Make sense.

File: core/freq_params_traffic/rule.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sczyh30) [reply core/freq_params_traffic/rule.go:21]: That's okay, but just `Kind` may be too generic. Maybe `ParamKind` or other concrete?

File: core/freq_params_traffic/concurrency_stat_slot.go (hunk 3 above)
Comment (louyuting) [reply core/freq_params_traffic/concurrency_stat_slot.go:25]: Okay

File: core/freq_params_traffic/traffic_shaping.go (hunk 2 above)
Comment (louyuting) [reply core/freq_params_traffic/traffic_shaping.go:220]: Okay

File: core/freq_params_traffic/rule.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (louyuting) [reply core/freq_params_traffic/rule.go:21]: ParamKind is Okay.

File: core/freq_params_traffic/slot.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+	if idx >= len(args) {
+		logger.Debugf("The argument doesn't exist for index(%d) of tc(%+v), args: %+v", idx, tc, args)
+		return nil
+	}
+	arg := args[idx]
+	if arg == nil {
+		return nil
+	}
+	switch arg.(type) {
+	case int:
```
Comment (sczyh30) [inline core/freq_params_traffic/slot.go:35]: Also support bool/float32/float64 here?

File: core/freq_params_traffic/traffic_shaping.go [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+	paramIndex    int
+	threshold     float64
+	specificItems map[interface{}]int64
+	durationInSec int64
+
+	metric *ParamsMetric
+}
+
+func newBaseTrafficShapingControllerWithMetric(r *Rule, metric *ParamsMetric) *baseTrafficShapingController {
+	size := int(math.Min(float64(ParamsMaxCapacity), float64(ParamsCapacityBase*r.DurationInSec)))
```
Comment (sczyh30) [inline core/freq_params_traffic/traffic_shaping.go:38]: How about making the cache size as an attribute of the param flow rule?

Comment (louyuting) [reply core/freq_params_traffic/traffic_shaping.go:38]: Maybe an optional attribute make sense.

**Comment by sczyh30** [comment]
I've been testing with the demo and found the pass count flaky like this (see the attachment):

[sentinel-go-demo-local-metrics.log](https://github.com/alibaba/sentinel-golang/files/4721561/sentinel-go-demo-local-metrics.log)

Rule:

```go
{
	Resource:          "test",
	MetricType:        freq_params_traffic.QPS,
	Behavior:          freq_params_traffic.Reject,
	ParamIndex:        0,
	Threshold:         100,
	MaxQueueingTimeMs: 0,
	ParamsMaxCapacity: 2,
	DurationInSec:     1,
	SpecificItems:     make(map[freq_params_traffic.SpecificValue]int64),
}
```

Demo:

```go
// Usually, the total pass QPS is expected to be 200 (two parameters, 100 respectively).
for i := 0; i < 8; i++ {
	go func() {
		for {
			e, b := sentinel.Entry(Resource, sentinel.WithSlotChain(sc), sentinel.WithArgs(true, rand.Int()%3000, uuid.New().String(), uuid.New().String()))
			if b != nil {
				// Blocked. We could get the block reason from the BlockError.
				time.Sleep(time.Duration(rand.Uint64()%10) * time.Millisecond)
			} else {
				// Passed, wrap the logic here.
				time.Sleep(time.Duration(rand.Uint64()%10) * time.Millisecond)
				// Be sure the entry is exited finally.
				e.Exit()
			}
		}
	}()
}

for {
	e, b := sentinel.Entry(Resource, sentinel.WithSlotChain(sc), sentinel.WithArgs(false, rand.Int()%3000, uuid.New().String(), uuid.New().String()))
	if b != nil {
		time.Sleep(time.Duration(rand.Uint64()%10) * time.Millisecond)
	} else {
		// Passed, wrap the logic here.
		time.Sleep(time.Duration(rand.Uint64()%10) * time.Millisecond)

		// Be sure the entry is exited finally.
		e.Exit()
	}
}
```

Any ideas about it?

**Comment by sczyh30** [comment]
Fabulous work. Thanks!

**Pull request merged**